File: stremio-addon/src/catalog.js
import fs from "node:fs";
import path from "node:path";
import { getTmdbMeta, getCachedMeta } from "./tmdb.js";

const SCRAPE_JSON = process.env.SCRAPE_JSON || "../moctale_items.json";

let lastMtime = 0;
let cachedItems = [];

function loadScrapedData() {
  const filePath = path.resolve(SCRAPE_JSON);
  try {
    const mtime = fs.statSync(filePath).mtimeMs;
    if (mtime === lastMtime && cachedItems.length) return cachedItems;

    const raw = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    const items = [];

    for (const [section, entries] of Object.entries(raw.sections)) {
      for (const entry of entries) {
        if (!entry.name) continue;
        items.push({ ...entry, section });
      }
    }

    lastMtime = mtime;
    cachedItems = items;
    return items;
  } catch {
    return cachedItems.length ? cachedItems : [];
  }
}

function inferType(name, year) {
  const showIndicators = /\b(season|episode|series|show)\b/i;
  const seriesNames = /tv\s*show|web\s*series|anime\s*series/i;
  if (showIndicators.test(name) || seriesNames.test(name)) return "series";
  if (year && parseInt(year, 10) <= new Date().getFullYear()) return "movie";
  return "movie";
}

function generateId(item) {
  if (item.link) {
    const slug = item.link.replace(/^https?:\/\/[^/]+\//, "").replace(/\/$/, "");
    if (slug) return `moctale-${slug}`;
  }
  const hash = Buffer.from(`${item.name}|${item.section}`).toString("base64url").slice(0, 16);
  return `moctale-${hash}`;
}

export async function getCatalog(type, genreFilter) {
  const items = loadScrapedData();
  const results = [];
  const filterLower = (genreFilter || "All").toLowerCase();

  for (const item of items) {
    const sectionLower = item.section.toLowerCase();

    if (filterLower !== "all") {
      const sectionSpaceless = sectionLower.replace(/\s+/g, "");
      const filterSpaceless = filterLower.replace(/\s+/g, "");
      if (filterSpaceless === "editorspick") {
        if (!sectionLower.includes("editor")) continue;
      } else if (!sectionSpaceless.includes(filterSpaceless) && !filterSpaceless.includes(sectionSpaceless)) {
        continue;
      }
    }

    const tmdb = await getTmdbMeta(item.name, item.year, item.type);
    const metaType = item.type || tmdb?.type || inferType(item.name, item.year);
    if (type !== metaType) continue;

    const descParts = [item.section];
    if (item.year) descParts.push(item.year);

    results.push({
      id: tmdb?.imdbId || generateId(item),
      type: metaType,
      name: item.name,
      poster: tmdb?.poster || item.poster_url,
      description: tmdb?.description || descParts.join(" · "),
      genres: tmdb?.genres || [],
      year: item.year || undefined,
    });
  }

  return results;
}

File: fastapi-addon/catalog.py
from __future__ import annotations

import base64
import re

from config import AddonConfig
from moctale import ScrapedItem, scrape_moctale
from tmdb import get_tmdb_meta


def infer_type(name: str, year: str) -> str:
    if re.search(r"\b(season|episode|series|show)\b", name, re.I):
        return "series"
    if year and int(year) <= 2100:
        return "movie"
    return "movie"


def section_matches_filter(section: str, genre_filter: str) -> bool:
    section_lower = section.lower()
    filter_lower = (genre_filter or "All").lower()

    if filter_lower == "all":
        return True

    section_flat = re.sub(r"\s+", "", section_lower)
    filter_flat = re.sub(r"\s+", "", filter_lower)

    if filter_flat == "editorspick":
        return "editor" in section_lower

    return filter_flat in section_flat or section_flat in filter_flat


def generate_id(item: ScrapedItem) -> str:
    if item.link:
        slug = item.link.rstrip("/").split("/")[-1]
        if slug:
            return f"moctale-{slug}"

    raw = f"{item.name}|{item.section}".encode("utf-8")
    digest = base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")[:16]
    return f"moctale-{digest}"


async def get_catalog(
    content_type: str,
    genre_filter: str,
    config: AddonConfig,
) -> tuple[list[dict], list[str]]:
    scrape = await scrape_moctale(config.moctale_cookie)
    metas: list[dict] = []

    for item in scrape["items"]:
        if not section_matches_filter(item.section, genre_filter):
            continue

        tmdb = await get_tmdb_meta(
            item.name,
            item.year,
            item.type,
            config.tmdb_api_key,
        )
        meta_type = item.type or (tmdb or {}).get("type") or infer_type(item.name, item.year)
        if meta_type != content_type:
            continue

        description_parts = [item.section]
        if item.year:
            description_parts.append(item.year)

        metas.append(
            {
                "id": (tmdb or {}).get("imdb_id") or generate_id(item),
                "type": meta_type,
                "name": item.name,
                "poster": (tmdb or {}).get("poster") or item.poster_url,
                "description": (tmdb or {}).get("description") or " - ".join(description_parts),
                "genres": (tmdb or {}).get("genres") or [],
                **({"year": item.year} if item.year else {}),
            }
        )

    return metas, scrape["notes"]
